backend/contracts/ExpenseTracker.sol

```diff
@@ -17,75 +18,90 @@ contract ExpenseTracker {
     }
 
     struct Tracker {
-        uint total;
-        uint deleted;
-        string[] transactions;
+        uint256 total;
+        uint256 deleted;
+        uint256 counter;
+        mapping(uint256 => string) transactions;
     }
-    mapping(address => mapping(string => Transaction)) private transactions;
+
     mapping(address => Tracker) private tracker;
     mapping(string => bool) private all_ids;
 
     function addExpense(
-        string memory _id,
         uint256 _amount,
         string memory _description,
         string memory _date,
         string memory _category
     ) public {
-        require(all_ids[_id] == false, "invalid id");
-        transactions[msg.sender][_id] = Transaction(
-            _id,
+        require(bytes(_category).length > 0, "category is required");
+        require(bytes(_date).length > 0, "date is required");
+
+        uint256 id = tracker[msg.sender].counter;
+        tracker[msg.sender].counter += 1;
+
+        Transaction memory expense = Transaction(
+            id,
             TransactionType.EXPENSE,
             _category,
             _description,
             _amount,
             _date,
             false
         );
+        
+        tracker[msg.sender].transactions[id] = _getTransactionId(id);
+        all_ids[_getTransactionId(id)] = true;
+
         tracker[msg.sender].total += 1;
-        tracker[msg.sender].transactions.push(_id);
-        all_ids[_id] = true;
     }
 
     function addIncome(
-        string memory _id,
         uint256 _amount,
         string memory _description,
         string memory _date,
         string memory _category
     ) public {
-        require(all_ids[_id] == false, "invalid id");
-        transactions[msg.sender][_id] = Transaction(
-            _id,
+        require(bytes(_category).length > 0, "category is required");
+        require(bytes(_date).length > 0, "date is required");

```

Comment:
Actually we must have a check that date is entered in correct format (YYYY-MM-DD), and it must not exceed current date.

Reply:
ok